reactive/containerd.py

```diff
@@ -257,21 +300,75 @@ def publish_config():
     )
 
 
-@when_not('containerd.br_netfilter.enabled')
-def enable_br_netfilter_module():
+@when('endpoint.docker-registry.ready')
+@when_not('containerd.registry.configured')
+def configure_registry():
     """
-    Enable br_netfilter to work
-    around https://github.com/kubernetes/kubernetes/issues/21613
+    Add docker registry config when present.
 
-    :returns: None
+    :return: None
     """
-    try:
-        modprobe('br_netfilter', persist=True)
-    except Exception:
-        log(traceback.format_exc())
-        if host.is_container():
-            log('LXD detected, ignoring failure to load br_netfilter')
-        else:
-            log('LXD not detected, will retry loading br_netfilter')
-            return
-    set_state('containerd.br_netfilter.enabled')
+    registry = endpoint_from_flag('endpoint.docker-registry.ready')
+
+    # Handle TLS data
+    cert_subdir = registry.registry_netloc
+    if registry.has_tls():
+        # Ensure the CA that signed our registry cert is trusted.
+        install_ca_cert(registry.tls_ca, name='juju-docker-registry')
+
+    manage_registry_certs(cert_subdir, remove=False)
+    # todo: add certs to config https://github.com/containerd/cri/blob/master/docs/registry.md
+
+    # Handle auth data.
+    if registry.has_auth_basic():
+        #  Set docker registry with auth.
+        docker_registry = {
+            "url": registry.registry_netloc,
+            "username": registry.basic_user,
+            "password": registry.basic_password
+        }
+    else:
+        #  Set docker registry without auth.
+        docker_registry = {"url": registry.registry_netloc}
+
+    DB.set('registry', docker_registry)
+
+    set_state('config.changed')
```

Comment:
This seems fishy to me since `config.changed` is normally handled by the config changed hook.  What happens after this is set, and could you have used different flags to trigger that effect without setting `config.changed`?

Reply:
Yeah, I would generally recommend not using `config.changed` like this but instead using a flag namespaced to the charm or layer specifically.  The two potential issues that come to mind are:

1) `config.changed` is cleared automatically (at the end of the hook), so you might not trigger the behavior you were hoping to
2) You might accidentally trigger other code that is watching for the `config.changed` flag

Neither of those are is terribly likely to cause you any problems here, but it seems easy enough to pick a different flag and change the handler you intend to trigger from: 

```python
@when('config.changed')
```

to

```python
@when_any('config.changed', 'containerd.registry.changed')
```

or even just replace the `set_state('config.changed')` call with an explicit call to `config_changed()` in this case.

And doing so would future-proof you against refactors that weren't aware of the `config.changed` flag overloading.